smartsim/entity/application.py

```diff
@@ -0,0 +1,300 @@
+        will be located in the path of the entity.  Invoking this method
+        after files have already been attached will overwrite
+        the previous list of entity files.
+
+        During generation, files "to_copy" are copied into
+        the path of the entity, and files "to_symlink" are
+        symlinked into the path of the entity.
+
+        Files "to_configure" are text based application input files where
+        parameters for the application are set. Note that only applications
+        support the "to_configure" field. These files must have
+        fields tagged that correspond to the values the user
+        would like to change. The tag is settable but defaults
+        to a semicolon e.g. THERMO = ;10;
+
+        :param to_copy: files to copy
+        :param to_symlink: files to symlink
+        :param to_configure: input files with tagged parameters
+        :raises ValueError: if the generator file already exists
+        """
+        to_copy = to_copy or []
+        to_symlink = to_symlink or []
+        to_configure = to_configure or []
+
+        # Check that no file collides with the parameter file written
+        # by Generator. We check the basename, even though it is more
+        # restrictive than what we need (but it avoids relative path issues)
+        for strategy in [to_copy, to_symlink, to_configure]:
+            if strategy is not None and any(
+                osp.basename(filename) == "smartsim_params.txt" for filename in strategy
+            ):
+                raise ValueError(
+                    "`smartsim_params.txt` is a file automatically "
+                    + "generated by SmartSim and cannot be ovewritten."
+                )
+        self.files = EntityFiles(to_configure, to_copy, to_symlink)
+
+    @property
+    def attached_files_table(self) -> str:
+        """Return a list of attached files as a plain text table
+
+        :returns: String version of table
+        """
+        if not self.files:
+            return "No file attached to this application."
+        return str(self.files)
+
+    def print_attached_files(self) -> None:
+        """Print a table of the attached files on std out"""
+        print(self.attached_files_table)
+
+    def __str__(self) -> str:  # pragma: no cover
+        entity_str = "Name: " + self.name + "\n"
+        entity_str += "Type: " + self.type + "\n"
+        entity_str += "Executable:\n"
+        for ex in self.exe:
+            entity_str += f"{ex}\n"
+        entity_str += "Executable Arguments:\n"
+        for ex_arg in self.exe_args:
+            entity_str += f"{str(ex_arg)}\n"
+        entity_str += f"Entity Files: {self.files}\n"
+        entity_str += f"File Parameters: {self.file_parameters}\n"
+        entity_str += "Incoming Entities:\n"
+        for entity in self.incoming_entities:
+            entity_str += f"{entity}\n"
+        entity_str += f"Key Prefixing Enabled: {self.key_prefixing_enabled}\n"
+
+        return entity_str
```

Comment:
If you want to avoid the manual string concatenation:
```suggestion
    def __str__(self) -> str:  # pragma: no cover
        exe_args_str = "\n".join(self.exe_args)
        entities_str = "\n".join(str(entity) for entity in self.incoming_entities)
        return textwrap.dedent(f"""\
            Name: {self.name}
            Type: {self.type}
            Executable:
            {self.exe}
            Executable Arguments:
            {exe_args_str}
            Entity Files: {self.files}
            File Parameters: {self.file_parameters}
            Incoming Entities:
            {entities_str}
            Key Prefixing Enabled: {self.key_prefixing_enabled}
            """)
```

Reply:
Also minor heads up: if two applications were to register each other and then one of them is cast to a `str`, this `__str__` impl will cause an infinite recursion and eventually stack overflow (or most likely `RecursionError`).

I don't think this was added as part of this PR so its not something we should address in this PR, but its probably something ticket and address later!

Reply:
Good point

Reply:
changed